processor/memorylimiter/memorylimiter.go

```diff
@@ -238,44 +239,89 @@ func (ml *memoryLimiter) forcingDrop() bool {
 	return atomic.LoadInt64(&ml.forceDrop) != 0
 }
 
-func (ml *memoryLimiter) memCheck() {
-	ms := ml.readMemStats()
-	ml.memLimiting(ms)
+func (ml *memoryLimiter) setForcingDrop(b bool) {
+	var i int64
+	if b {
+		i = 1
+	} else {
+		i = 0
+	}
+	atomic.StoreInt64(&ml.forceDrop, i)
 }
 
-func (ml *memoryLimiter) memLimiting(ms *runtime.MemStats) {
-	if !ml.decision.shouldDrop(ms) {
-		atomic.StoreInt64(&ml.forceDrop, 0)
-	} else {
-		atomic.StoreInt64(&ml.forceDrop, 1)
-		// Force a GC at this point and see if this is enough to get to
-		// the desired level.
+func bytesToMib(ms *runtime.MemStats) zap.Field {
+	return zap.Uint64("cur_mem_mib", ms.Alloc/1024/1024)
+}
+
+func (ml *memoryLimiter) checkMemLimits() {
+	ms := ml.readMemStats()
+
+	ml.logger.Debug("Currently used memory.", bytesToMib(ms))
+
+	if ml.usageChecker.aboveHardLimit(ms) {
+		// Memory usage is above hard limit. Force GC.
+		ml.logger.Warn("Memory usage is above hard limit, force GC.", bytesToMib(ms))
 		runtime.GC()
+		ms = ml.readMemStats()
+		ml.logger.Info("Memory usage after GC.", bytesToMib(ms))
+	}
```

Comment:
before GC-ing I would set the dropping bit in case it was a jump from the non-dropping (less than soft) to greater than hard.

Reply:
I don't want to begin dropping sooner than we have to. GC has a good chance to free memory and avoid starting dropping. I would prefer to keep the logic this way.
In fact, I am thinking to invoke GC even when we pass the soft limit for the first time in the upward direction. Probably can be improved in a future PR.

Reply:
If you don't drop you have a big chance of OOM-ing at this moment, you are above the hard limit....

Reply:
This should not happen if interval and spike settings are selected correctly. If they are selected incorrectly nothing will help.

Reply:
Updated to do GC earlier.